crates/bsp_types/src/extensions/rust/rust_workspace.rs

```diff
@@ -16,141 +16,236 @@ impl Request for RustWorkspace {
     #[serde(skip_serializing_if = "Option::is_none")]
     pub kind: Option<String>,
+    /** The target platform for the dependency. */
     #[serde(skip_serializing_if = "Option::is_none")]
     pub target: Option<String>,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub optional: Option<bool>,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub uses_default_features: Option<bool>,
-    pub features: Vec<String>,
+    /** Indicates whether this is an optional dependency. */
+    pub optional: bool,
+    /** Indicates whether default features are enabled. */
+    pub uses_default_features: bool,
+    /** A sequence of enabled features. **/
+    pub features: BTreeSet<String>,
 }
 
-#[derive(Serialize, Deserialize, Default)]
+/** This structure is embedded in the `data?: BuildTargetData` field, when the
+`dataKind` field contains "rust". */
+#[derive(Serialize, Deserialize, Default, Clone)]
 #[serde(rename_all = "camelCase")]
-pub struct RustTarget {
+pub struct RustBuildTarget {
+    /** The name of the target. */
     pub name: String,
-    pub crate_root_url: String,
-    pub package_root_url: String,
+    /** Path to the root module of the crate. */
+    pub crate_root_url: Uri,
+    /** A target's kind. */
     pub kind: RustTargetKind,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub edition: Option<String>,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub doctest: Option<bool>,
-    #[serde(default, skip_serializing_if = "Vec::is_empty")]
-    pub required_features: Vec<String>,
+    /** Type of output that is produced by a crate during the build process.
+    The crate type determines how the source code is compiled. */
+    pub crate_types: Vec<RustCrateType>,
+    /** The Rust edition of the target. */
+    pub edition: Edition,
+    /** Whether or not this target has doc tests enabled, and
+    the target is compatible with doc testing. */
+    pub doctest: bool,
+    /** A sequence of required features. */
+    pub required_features: BTreeSet<String>,
 }
 
-#[derive(Serialize_repr, Deserialize_repr, Default)]
+#[derive(Serialize_repr, Deserialize_repr, Default, Clone)]
 #[repr(u8)]
 pub enum RustTargetKind {
+    /** For lib targets. */
     #[default]
     Lib = 1,
+    /** For binaries. */
     Bin = 2,
+    /** For integration tests. */
     Test = 3,
+    /** For examples. */
     Example = 4,
+    /** For benchmarks. */
     Bench = 5,
+    /** For build scripts. */
     CustomBuild = 6,
+    /** For unknown targets. */
     Unknown = 7,
 }
 
+/** Crate types (`lib`, `rlib`, `dylib`, `cdylib`, `staticlib`) are listed for
+`lib` and `example` target kinds. For other target kinds `bin` crate type is listed. */
+#[derive(Serialize_repr, Deserialize_repr, Default, Clone)]
+#[repr(u8)]
+pub enum RustCrateType {
+    Bin = 1,
+    #[default]
+    Lib = 2,
+    Rlib = 3,
+    Dylib = 4,
+    Cdylib = 5,
+    Staticlib = 6,
+    ProcMacro = 7,
+    Unknown = 8,
+}
```

Comment:
czemu te enumy to intEnumy? :(

Reply:
Andrzej mi kiedyś powiedział, że ogólna konwencja jest taka, że jeśli wartości enuma są znane z góry (czyli raczej się nie będą zmieniać), to enum jest liczbowy i tego się trzymam

Reply:
no ale tu mogą być jakieś zmiany chyba? No i to chyyyyyba powinien być open enum który tym bardziej lepiej żeby nie był liczbowy... Ale no jak Andrzej twierdzi że tak jest powinno być to okay, on ma decydujący głos

Reply:
Musiałyby zapewne powstać jakieś głębokie zmiany w cargo, żeby to się zmieniło. A sugerowałam się wcześniejszymi uwagami do PR drugiego teamu, choć cały PR nie jest jeszcze zatwierdzony. Więc jeśli jednak zostanie uznane inaczej to to zmienię i tutaj

Reply:
git